### PR TITLE
[Model] Simplify RGCN

### PR DESCRIPTION
**Pull request by mufeili**
## Description
- Remove the logic for sampling-based RGCN on ogbn-mag as there have been separate examples for it
- Unify the RGCN model definition across the three settings (entity classification, sampling-based entity classification, link prediction)
- Split `entity_classify_mp.py` into `entity_sample.py` and `entity_sample_multi_gpu.py`, one for single-gpu and one for multi-gpu
- Rewrite the subgraph sampler and iterator for link prediction in `link_utils.py`, which might be of interest to @BarclayII 
- Other simplification for better readability
- Update the README file with new instructions and performance numbers
- Fix a minor error in the doc of `FB15k237Dataset`

@jermainewang , I haven't used an external package for computing mean reciprocal rank in link prediction as it seems to be quite intuitive in this case. I've done some investigation on this:
- the one implemented in TorchMetrics only works for binary classification
- the one implemented in OGB only works for a pre-defined set of datasets in initializing `Evaluator`
- `label_ranking_average_precision_score` from sklearn seems to work

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## TODO in separate PRs
- Move multi-gpu sampling-based entity classification to a separate example
- Try implementing multi-gpu training with PyTorch Lightning

### CODE REVIEW

File: examples/pytorch/rgcn/model.py [hunk 1]
```diff
@@ -1,166 +1,105 @@
...
+from dgl.nn.pytorch import RelGraphConv
 
-        # create rgcn layers
-        self.build_model()
+class RGCN(nn.Module):
+    """
+    Parameters
+    ----------
+    in_dim : int
+        Input feature size or number of nodes
```
Comment (jermainewang) [inline examples/pytorch/rgcn/model.py:14]: It's ok if you don't want to have very long docstring in examples as long as the argument name is self-explaining, but don't leave something half done.

File: examples/pytorch/rgcn/entity.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+def main(args):
+    g, num_rels, num_classes, labels, train_idx, test_idx, target_idx = load_data(
+        args.dataset, get_norm=True)
+
+    num_nodes = g.num_nodes()
+
+    # Since the nodes are featureless, learn node embeddings from scratch
+    # This requires passing the node IDs to the model.
+    feats = th.arange(num_nodes)
```
Comment (jermainewang) [inline examples/pytorch/rgcn/entity.py:24]: Maybe we shouldn't depend on this behavior anymore. How about we directly create an embedding tensor? Is it equivalent?

File: examples/pytorch/rgcn/entity_sample.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    # map the seed nodes back to their type-specific ids,
+    # in order to get the target node labels
+    seeds = inv_target[seeds]
+
+    for blc in blocks:
+        _, v, eid = blc.edges(form='all')
+        _, inverse_index, count = th.unique(v, return_inverse=True, return_counts=True)
+        degrees = count[inverse_index]
+        norm = th.ones(eid.shape[0]).to(degrees.device) / degrees
+        blc.edata['norm'] = norm.unsqueeze(1)
```
Comment (jermainewang) [inline examples/pytorch/rgcn/entity_sample.py:83]: Is this normalizer calculation a common thing? Shall we add it as a DGL utility?

File: examples/pytorch/rgcn/entity_sample.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+    return seeds, blocks
+
+def train(model, embed_layer, train_loader, inv_target,
+          labels, emb_optimizer, optimizer):
+    model.train()
+    embed_layer.train()
+
+    for sample_data in train_loader:
+        seeds, blocks = process_batch(inv_target, sample_data)
```
Comment (jermainewang) [inline examples/pytorch/rgcn/entity_sample.py:93]: It is difficult to decipher what this `inv_target` is for from the context.

Comment (jermainewang) [reply examples/pytorch/rgcn/entity_sample.py:93]: It is generated from `load_data` utility so I need to look up there. Unlike the other arrays, I cannot understand it just by its name.

File: examples/pytorch/rgcn/entity_sample.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+    return eval_logits, eval_seeds
+
+def main(args):
+    g, num_rels, num_classes, labels, train_idx, test_idx, target_idx, inv_target = load_data(
+        args.dataset, inv_target=True)
+
+    # Create csr/coo/csc formats before launching training processes.
+    # This avoids creating certain formats in each data loader process, which saves momory and CPU.
+    g.create_formats_()
```
Comment (jermainewang) [inline examples/pytorch/rgcn/entity_sample.py:133]: Is this handled by dataloader?

File: examples/pytorch/rgcn/entity_sample.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='RGCN for entity classification with sampling')
+    parser.add_argument("--dropout", type=float, default=0,
+                        help="dropout probability")
+    parser.add_argument("--n-hidden", type=int, default=16,
+                        help="number of hidden units")
+    parser.add_argument("--gpu", type=int, default=0,
+                        help="gpu")
+    parser.add_argument("--sparse-lr", type=float, default=2e-2,
+                        help="sparse embedding learning rate")
```
Comment (jermainewang) [inline examples/pytorch/rgcn/entity_sample.py:178]: Shouldn't it use the same lr?

File: examples/pytorch/rgcn/entity_sample_multi_gpu.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+from entity_utils import load_data
+from entity_sample import init_dataloaders, init_models, train, evaluate
+
+def run(proc_id, n_gpus, n_cpus, args, devices, dataset, queue=None):
+    dev_id = devices[proc_id]
+    g, num_classes, num_rels, target_idx, inv_target, train_idx,\
+        test_idx, labels = dataset
+
+    dist_init_method = 'tcp://{master_ip}:{master_port}'.format(
+        master_ip='127.0.0.1', master_port='12345')
```
Comment (jermainewang) [inline examples/pytorch/rgcn/entity_sample_multi_gpu.py:25]: Is this a standard practice?

File: examples/pytorch/rgcn/entity_sample_multi_gpu.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    backend = 'nccl'
+
+    if not args.dgl_sparse:
+        backend = 'gloo'
```
Comment (jermainewang) [inline examples/pytorch/rgcn/entity_sample_multi_gpu.py:29]: `backend = 'nccl' if args.dgl_sparse else 'gloo'`

Try avoid negation in condition to increase readability.

File: examples/pytorch/rgcn/entity_sample_multi_gpu.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    use_ddp = True if n_gpus > 1 else False
+    train_loader, val_loader, test_loader = init_dataloaders(
+        args, g, train_idx, test_idx, target_idx, dev_id, use_ddp=use_ddp)
+    embed_layer, model = init_models(args, device, g.num_nodes(), num_classes, num_rels)
+
+    labels = labels.to(device)
+    model = model.to(device)
+    model = DistributedDataParallel(model, device_ids=[dev_id], output_device=dev_id)
+    if not args.dgl_sparse:
+        embed_layer = DistributedDataParallel(embed_layer, device_ids=None, output_device=None)
```
Comment (jermainewang) [inline examples/pytorch/rgcn/entity_sample_multi_gpu.py:47]: Generally speaking, I feel dgl_sparse introduces quite a lot complexity into the example code. Shall we remove it from the example code entirely? @classicsong What's your opinion?

File: examples/pytorch/rgcn/entity_sample_multi_gpu.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    for epoch in range(args.n_epochs):
+        train_loader.set_epoch(epoch)
+        train_acc, loss = train(model, embed_layer, train_loader, inv_target,
+                                labels, emb_optimizer, optimizer)
+
+        if proc_id == 0:
+            print("Epoch {:05d}/{:05d} | Train Accuracy: {:.4f} | Train Loss: {:.4f}".format(
+                epoch, args.n_epochs, train_acc, loss))
+
+        gc.collect()
```
Comment (jermainewang) [inline examples/pytorch/rgcn/entity_sample_multi_gpu.py:79]: Better remove it or add an inline comment to explain.

File: examples/pytorch/rgcn/entity_sample_multi_gpu.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+        embed_layer = DistributedDataParallel(embed_layer, device_ids=None, output_device=None)
+
+    if args.dgl_sparse:
+        emb_optimizer = dgl.optim.SparseAdam(params=[embed_layer.node_embed],
+                                             lr=args.sparse_lr, eps=1e-8)
+    else:
+        emb_optimizer = th.optim.SparseAdam(embed_layer.module.parameters(), lr=args.sparse_lr)
+    optimizer = th.optim.Adam(model.parameters(), lr=1e-2, weight_decay=args.l2norm)
+
+    def collect_eval():
```
Comment (jermainewang) [inline examples/pytorch/rgcn/entity_sample_multi_gpu.py:56]: Does it need to be a nested function?

File: examples/pytorch/rgcn/entity_utils.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    # find out the target node ids in g
+    loc = (g.ndata['ntype'] == category_id)
+    target_idx = node_ids[loc]
+
+    if inv_target:
+        # Map global node IDs to type-specific node IDs. This is required for
+        # looking up type-specific labels in a minibatch
+        inv_target = th.empty((g.num_nodes(),), dtype=th.int64)
+        inv_target[target_idx] = th.arange(0, target_idx.shape[0],
+                                           dtype=inv_target.dtype)
```
Comment (jermainewang) [inline examples/pytorch/rgcn/entity_utils.py:59]: I wonder what is the best way to avoid this. Could you use the previously saved `g.ndata['type_id']`?

File: examples/pytorch/rgcn/link_utils.py [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+https://github.com/MichSchli/RelationPrediction
+
+"""
+import numpy as np
+import torch as th
+import dgl
+
+# Utility function for building training and testing graphs
+
+def comp_deg_norm(g):
```
Comment (jermainewang) [inline examples/pytorch/rgcn/link_utils.py:13]: I feel this is also quite common. I could definitely see Transform module should include it.

File: examples/pytorch/rgcn/link_utils.py [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+    if bidirected:
+        sub_src, sub_dst = th.cat([sub_src, sub_dst]), th.cat([sub_dst, sub_src])
+        sub_rel = th.cat([sub_rel, sub_rel + num_rels])
+
+    sub_g = dgl.graph((sub_src, sub_dst), num_nodes=g.num_nodes())
+    sub_g.edata[dgl.ETYPE] = sub_rel
+
+    return sub_g
+
+def preprocess(g, num_rels):
```
Comment (jermainewang) [inline examples/pytorch/rgcn/link_utils.py:36]: @VoVAllen Probably this could be the starting point for the `split_edges` utility.

**Review by jermainewang — COMMENTED** [review]
Left some comments. Skipped `link.py` as I feel we will back to it once the sampling refactoring is done.

File: examples/pytorch/rgcn/model.py (hunk 1 above)
Comment (mufeili) [reply examples/pytorch/rgcn/model.py]: Removed.

File: examples/pytorch/rgcn/entity.py (hunk 2 above)
Comment (mufeili) [reply examples/pytorch/rgcn/entity.py:24]: Mathematically, it is not equivalent primarily due to the basis-decomposition in initializing node embeddings. Meanwhile, it's possible to initialize a learnable embedding per node without basis decomposition as in the implementation of sampling-based entity classification. However, it will require additional efforts to tune the hyperparameters to see if we can match the performance numbers.

File: examples/pytorch/rgcn/entity_sample.py (hunk 3 above)
Comment (mufeili) [reply examples/pytorch/rgcn/entity_sample.py]: Added `norm_by_dst` in `transform.py`.

File: examples/pytorch/rgcn/entity_sample.py (hunk 4 above)
Comment (mufeili) [reply examples/pytorch/rgcn/entity_sample.py:90]: There is a comment in `process_batch` that explains what `inv_target` is. Any ideas for improvement?

File: examples/pytorch/rgcn/entity_sample.py (hunk 6 above)
Comment (mufeili) [reply examples/pytorch/rgcn/entity_sample.py:175]: I can try performing a hyperparameter tuning to see if a same lr can work.

File: examples/pytorch/rgcn/entity_sample.py (hunk 5 above)
Comment (mufeili) [reply examples/pytorch/rgcn/entity_sample.py:130]: Removed.

File: examples/pytorch/rgcn/entity_sample_multi_gpu.py (hunk 8 above)
Comment (mufeili) [reply examples/pytorch/rgcn/entity_sample_multi_gpu.py:29]: Done.

File: examples/pytorch/rgcn/entity_sample_multi_gpu.py (hunk 7 above)
Comment (mufeili) [reply examples/pytorch/rgcn/entity_sample_multi_gpu.py:25]: PyTorch's doc simply used `'tcp://10.1.1.20:23456'` for tcp initialization as an example [here](https://pytorch.org/docs/stable/distributed.html#tcp-initialization).

File: examples/pytorch/rgcn/entity_sample_multi_gpu.py (hunk 11 above)
Comment (mufeili) [reply examples/pytorch/rgcn/entity_sample_multi_gpu.py]: Changed to be non-nested.

File: examples/pytorch/rgcn/entity_utils.py (hunk 12 above)
Comment (mufeili) [reply examples/pytorch/rgcn/entity_utils.py:55]: You are right. They are equivalent. Then do you prefer passing `g` to `train` and `evaluate` instead and accessing `g.ndata['type_id']` within them?

File: examples/pytorch/rgcn/link_utils.py (hunk 13 above)
Comment (mufeili) [reply examples/pytorch/rgcn/link_utils.py:13]: They are actually equivalent. Replaced.

**Comment by mufeili** [comment]
I also noticed that by removing `.long()` after `dgl.norm_by_dst(g)` in `link_utils.py`, the MRR significantly increased, from 0.18 to 0.39 for the raw MRR and from 0.26 to 0.56 for the filtered MRR. `.long()` casts float values like 0.33 to 0, which seems to be a bug.

File: examples/pytorch/rgcn/link.py [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+        optimizer.step()
+
+        print("Epoch {:04d} | Loss {:.4f} | Best MRR {:.4f}".format(epoch, loss.item(), best_mrr))
+
+        if (epoch + 1) % 500 == 0:
+            # perform validation on CPU because full graph is too large
+            model = model.cpu()
+            model.eval()
+            print("start eval")
+            embed = model(test_g, test_node_id)
```
[the remark this answers is not in the document]
Comment (mufeili) [reply examples/pytorch/rgcn/link.py:100]: Nice catch! I've fixed it accordingly.

**Comment by mufeili** [comment]
@jermainewang I've removed code snippets depending on dgl sparse. Since this PR also involves code changes to DGL core, it will be most ideal to merge it before code freeze.

**Review by jermainewang — APPROVED** [review]
